go/obsclient/authclient.go

```diff
@@ -0,0 +1,78 @@
+package obsclient
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/obscuronet/go-obscuro/go/rpcclientlib"
+)
+
+// AuthObsClient extends the functionality of the ObsClient for all methods that require encryption when communicating with the enclave
+type AuthObsClient struct {
+	ObsClient
+	c rpcclientlib.Client
+}
+
+// NewAuthObsClient constructs an AuthObsClient for sensitive communication with an enclave.
+//
+//	It requires an EncRPCClient specifically even though the AuthObsClient uses a Client interface in its struct because
+//	the Client interface makes testing easy but an EncRPCClient is required for the actual encrypted communication
+func NewAuthObsClient(client *rpcclientlib.EncRPCClient) *AuthObsClient {
+	return &AuthObsClient{
+		ObsClient: ObsClient{
+			c: client,
+		},
+		c: client,
+	}
+}
+
+func (ac *AuthObsClient) Close() {
+	ac.c.Stop()
+	ac.ObsClient.Close()
+}
+
+// TransactionByHash returns transaction (if found), isPending (always false currently as we don't search the mempool), error
+func (ac *AuthObsClient) TransactionByHash(ctx context.Context, hash common.Hash) (*types.Transaction, bool, error) {
+	var tx types.Transaction
+	err := ac.c.CallContext(ctx, &tx, rpcclientlib.RPCGetTransactionByHash, hash.Hex())
+	// todo: revisit isPending result value, included for ethclient equivalence but hardcoded currently
+	return &tx, false, err
+}
+
+func (ac *AuthObsClient) TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
+	var receipt types.Receipt
+	err := ac.c.CallContext(ctx, &receipt, rpcclientlib.RPCGetTxReceipt, txHash)
+	return &receipt, err
+}
+
+func (ac *AuthObsClient) NonceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (uint64, error) {
+	var result hexutil.Uint64
+	err := ac.c.CallContext(ctx, &result, rpcclientlib.RPCNonce, account, toBlockNumArg(blockNumber))
```

Comment:
Shouldn't this block number conversion happen at the level of the actual client (e.g. EncRPCClient), not at the authentication layer? Feels like this layer should just be pass through, except for authentication issues as needed.

Reply:
This client is for translation from Go objects to RPC parameters so I think this is the right place for this sort of stuff (this same conversion happens in the EthClient).